lib/std/crypto/argon2.zig

```diff
@@ -0,0 +1,1119 @@
+) void {
+    const lanes = memory / threads;
+    var lane: u32 = 0;
+    while (lane < threads - 1) : (lane += 1) {
+        for (blocks.items[(lane * lanes) + lanes - 1]) |v, i| {
+            blocks.items[memory - 1][i] ^= v;
+        }
+    }
+    var block: [1024]u8 = undefined;
+    for (blocks.items[memory - 1]) |v, i| {
+        mem.writeIntLittle(u64, block[i * 8 ..][0..8], v);
+    }
+    switch (out.len) {
+        Blake2b512.digest_length => blake2bHash(out, &block, Blake2b512),
+        Blake2b384.digest_length => blake2bHash(out, &block, Blake2b384),
+        Blake2b256.digest_length => blake2bHash(out, &block, Blake2b256),
+        Blake2b192.digest_length => blake2bHash(out, &block, Blake2b192),
+        Blake2b160.digest_length => blake2bHash(out, &block, Blake2b160),
+        Blake2b128.digest_length => blake2bHash(out, &block, Blake2b128),
+        else => {
+            if (out.len % 32 == 0) {
+                blake2bHash(out, &block, Blake2b512);
+            } else if (hasher) |h| {
+                blake2bHash(out, &block, h);
+            } else {
+                unreachable;
+            }
+        },
+    }
+}
+
+fn indexAlpha(
+    rand: u64,
+    lanes: u32,
+    segments: u32,
+    threads: u32,
+    n: u32,
+    slice: u32,
+    lane: u32,
+    index: u32,
+) u32 {
+    var ref_lane = @intCast(u32, rand >> 32) % threads;
+    if (n == 0 and slice == 0) {
+        ref_lane = lane;
+    }
+    var m = 3 * segments;
+    var s = ((slice + 1) % sync_points) * segments;
+    if (lane == ref_lane) {
+        m += index;
+    }
+    if (n == 0) {
+        m = slice * segments;
+        s = 0;
+        if (slice == 0 or lane == ref_lane) {
+            m += index;
+        }
+    }
+    if (index == 0 or lane == ref_lane) {
+        m -= 1;
+    }
+    return phi(rand, m, s, ref_lane, lanes);
+}
+
+fn phi(
+    rand: u64,
+    m: u64,
+    s: u64,
+    lane: u32,
+    lanes: u32,
+) u32 {
+    var p = rand & 0xffff_ffff;
+    p = (p * p) >> 32;
+    p = (p * m) >> 32;
+    return lane * lanes + @intCast(u32, ((s + m - (p + 1)) % lanes));
+}
+
+/// Derives a key from the password, salt, and argon2 parameters.
+///
+/// Derived key [l]ength has to be in 4 <= l <= 64 or l % 32 == 0.
+///
+/// Salt has to be at least 8 bytes legth.
+///
+/// The [hasher] is Blake2b(derived_key.len * 8). It is required when derived_key [l]ength 
+/// not in [16, 20, 24, 32, 48, 64] and l < 64.
+pub fn kdf(
+    allocator: *mem.Allocator,
+    derived_key: []u8,
+    password: []const u8,
+    salt: []const u8,
+    params: Params,
+    mode: Mode,
+    comptime hasher: ?type,
+) KdfError!void {
+    if (derived_key.len < 4 or derived_key.len > max_int) return KdfError.OutputTooLong;
+    if (password.len > max_int) return KdfError.WeakParameters;
+    if (salt.len < 8 or salt.len > max_int) return KdfError.WeakParameters;
+    if (params.t < 1 or params.p < 1) return KdfError.WeakParameters;
+
+    if (hasher == null and
```

Comment:
`hasher` is not necessary.

You can use `Blake2b(512).init(.expected_out_bits = out_len)` and then truncate the output to `out_len` bytes.